rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,856 @@
+
+- `ctd-ecs-app` for ECS service and load balancer configuration
+- `ctd-rds-db` for RDS database configuration
+
+These _infrastructure component modules_ have seen some adoption, but are not yet widely used. To maximize their effectiveness, we will:
+
+- Set up a testing workflow for each module using the process described in the [Testing Workflow][guide-module-testing] section
+- Move them to separate repos and version them, using the process described in the [Version and Publish Modules section][guide-module-versioning]
+
+We will also push to migrate all applications to use these modules to minimize redundant and nonstandard configuration.
+
+Relatedly, the `ctd-ecs-app` module introduced the concept of a `context` var that bundles all baseline configuration need by a module, including VPC and subnet information, and shared resources such as TLS certificates and Cognito domains. This convention has the potential to significantly reduce redundant configuration in app modules, and should also be adopted widely.
+
+See the corresponding [reference-level explanation][reference-standardization] for related recommendations.
+
+## 4. Version and Publish Modules
+[guide-module-versioning]: #4-version-and-publish-modules
+
+_Modules as immutable artifacts_
+
+Nearly all of CTD's modules are currently stored in the devops repo and called using a relative path to the module's directory. Terraform supports a number of alternate methods for sourcing modules, including:
+
+- Git repositories (including specific refs)
+- Public module registries
+- Private module registries
+
+The main advantage of using a repo or module registry instead of a path is that the module can then be treated as a versioned artifact. If we are able to pin module calls to a specific version, we can make changes to the module without immediately impacting all the contexts in which it's called. This is useful for both infrastructure component modules, which are used across many different applications, and application modules, which are used to build multiple distinct environments. It also reduces the need to rely on feature gating mechanisms such as module variables, which is the typical approach to module changes currently.
+
+More importantly, versioning allows us to explicitly declare what version of an application module is running in the production context, and enforce a change control process to ensure that no unwanted or unintended changes are applied. Change control on production systems is a security requirement, which CTD typically handles via code reviews or deployment approvals. Without module versioning, changes to child modules could have immediate and undetectable impacts to production systems. By pinning modules in production, we can isolate changes to each running environment, and thus isolate the approval step for an extra layer of protection.
```

Comment:
Re: changes being undetectable, wouldn't they show up in `terraform plan`?

Reply:
Changes to child modules will show up in a `terraform plan` in the module where that child module is called, yes. The problem is that, when merging changes to a child module, there's no way to know where it's being called, and thus no way to know whether or not production resources are being impacted by the change.

If we're running `terraform plan` and `terraform apply` in an automated manner (as is proposed elsewhere in this RFC), then we need to be sure that **every** plan and apply contains only the changes we desire. There is no way to guarantee that without versioning child modules in production.